src/main/java/com/iexec/sms/secret/web2/Web2SecretsService.java

```diff
@@ -48,27 +48,15 @@ public Optional<Secret> getSecret(String ownerAddress, String secretAddress) {
 
     public Optional<Secret> getSecret(String ownerAddress, String secretAddress, boolean shouldDecryptValue) {
         ownerAddress = ownerAddress.toLowerCase();
-        Optional<Web2Secrets> web2Secrets = getWeb2Secrets(ownerAddress);
-        if (web2Secrets.isEmpty()) {
-            return Optional.empty();
-        }
-        Secret secret = web2Secrets.get().getSecret(secretAddress);
-        if (secret == null) {
-            return Optional.empty();
-        }
-        if (shouldDecryptValue) {
-            decryptSecret(secret);
-        }
-        return Optional.of(secret);
+        Optional<Secret> oSecret = getWeb2Secrets(ownerAddress)
+                .flatMap(web2Secrets -> web2Secrets.getSecret(secretAddress));
+        return shouldDecryptValue ? oSecret.map(this::decryptSecret) : oSecret;
     }
 
     public void addSecret(String ownerAddress, String secretAddress, String secretValue) {
         ownerAddress = ownerAddress.toLowerCase();
-        Web2Secrets web2Secrets = new Web2Secrets(ownerAddress);
-        Optional<Web2Secrets> existingWeb2Secrets = getWeb2Secrets(ownerAddress);
-        if (existingWeb2Secrets.isPresent()) {
-            web2Secrets = existingWeb2Secrets.get();
-        }
+        Web2Secrets web2Secrets = getWeb2Secrets(ownerAddress)
```

Comment:
Not sure to understand why we authorize to "re-"add a secret (?)
I feel we should return as soon as possible if a secret is already set here

Reply:
That's pretty strange, you're right. It was the original operation but it may change. What about using a `Map` instead of a list for `Web2Secrets#secrets` so that it is easier to look for a particular secret? Then, we can update `addSecret` to check the currently-adding secret is not already known. Finally, return a boolean to tell whether the operation succeeded.
BTW, this check is already done in `SecretController#addWeb2Secret` but I'm not sure this is its place.

Reply:
I think the check you're asking for is here
https://github.com/iExecBlockchainComputing/iexec-sms/blob/feature/improve-web2-and-web3-secrets-coverage/src/main/java/com/iexec/sms/secret/SecretController.java#L138

The secret won't be added a second time if it already exists.

If it already exists, 2 things can happen:
* It is the first Web2Secrets and it must be created (the line the comment is pointing at)
* It is not the first Web2Secrets, we must fetch Web2Secrets, access its secrets field and append a new Secret.

The key point here is that a `Web2Secrets` is not a web2 secret, it is a wrapper around the web2 secrets collection of an owner.

Reply:
Line were actual secret is appended:
https://github.com/iExecBlockchainComputing/iexec-sms/blob/feature/improve-web2-and-web3-secrets-coverage/src/main/java/com/iexec/sms/secret/web2/Web2SecretsService.java#L65

Reply:
I feel we should check in `addSecret()` that no secret already exists (then adding a `return;` condition if it is the case) but then use a fresh `new Web2Secrets(ownerAddress)`

Reply:
see 916c786